writer.go

```diff
@@ -957,6 +966,24 @@ func (w *Writer) Stats() WriterStats {
 	}
 }
 
+func (w *Writer) chooseTopic(msg Message) (string, error) {
+	if w.Topic != "" {
+		if msg.Topic != "" {
+			// if writer topic defined, messages must not also specify one
+			return "", InvalidMessage
+		}
+
+		return w.Topic, nil
+	} else {
+		if msg.Topic == "" {
+			// either the writer or the message must specify a topic
+			return "", InvalidTopic
+		}
+
+		return msg.Topic, nil
+	}
```

Comment:
Maybe this function can be expressed with a bit less branching to help readers grasp the logc?

```go
if (w.Topic != "" && msg.Topic != "") || (w.Topic == "" && msg.Topic == "") {
    return "", InvalidMessage
}
if msg.Topic != "" {
    return msg.Topic, nil
}
return w.Topic, nil
```
This could be a personal preference tho, if the form looks harder to read to you it's probably fine to keep it the way it is.

Reply:
Ah yeah, I was going back and forth with various ways to accomplish this, this looks more concise so I'll adopt it.